### ISSUE TITLE
List of SAML TRs doesn't show the actual status of its displayed items.

### ISSUE THREAD

**Issue by aliaksander-samuseu** [GluuFederation/oxTrust#385]
Environment:
CentOS 6.7, Gluu CE 2.4.4 SP2

Preconditions:
Gluu was installed with Shibboleth module

Steps to reproduce:
1. Log in to web UI as admin user
2. Move to "SAML -> Trust Relationships" page and click "Add TR" button there
3. Fill in all fields on the page required to create some simple valid TR there, using either "File" or "Url" methods. Make sure metadata you'll use is valid/reachable, i.e. under normal circumstances it must pass validation successfully in Gluu and result in active TR.
4. Click the "Update" button to finalize the creation. In case of success you'll still should be at the "Update Trust Relationships" page displaying TR you've just created
5. Click the "Trust Relationships" sub-menu item in the accordion menu to the left to get to the list of all TRs again. It's very important to click this very item first, i.e. you must not click some other item in the menu, and only after that this one, this is not guaranteed to reproduce the issue!
6. Click the "Search" button (leaving the search field empty) right away to display all TRs. Do not leave this page any more, just try to click the "Search" button again and again for several minutes (what usually is enough for any simple metadata to pass validation), observing status of TR you created before.

Result:
After hitting the button on step 6) you'll see that your TR is not yet validated/active. Now even if you'll stay on this page for 5 minutes, clicking it, you'll still see its state doesn't change. Simply refreshing the page (by hitting F5/Ctrl+R) won't help either. The only way to get it correctly updated is to actually leave to some other page (OIDC, Asimba etc), then returning to this one again. You also can check corresponding attributes of this TR in LDAP after 1-2 mins, you'll see they already contain updated states, while in the list it still isn't updated. Please see attached screenshot.

Please note that it may be mandatory to complete step 5) very quickly, so at the moment you'll click "Search" button on step 6) your TR still will hasn't been validated/activated. If you have a top-notch machine in your disposal, I guess it may pass validation very quickly.

**It's quite possible that similar issue may be present on other oxTrust's pages with a similar functionality too.**

Expected result:
The page should display actual status of all items without the need to leave it and return again, simply hitting the "Search" button to refresh the list and/or reloading the page the usual way should be enough.

![tr_list_no_updates](https://cloud.githubusercontent.com/assets/11802046/20034131/9651a13a-a3c4-11e6-912f-98df8b36cafa.jpg)

**Comment by shekhar16** [comment]
@aliaksander-samuseu   to refresh the the search result you need to change the search pattern
suppose if you had "test" keyword in text box, to get updated result you need to change the text in textbox.otherwise it will not hit the ldap and show the old result

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

You can be one 100% sure most of users won't get it that easily :) As always to get the list they simply click that button with empty field, that's what they will continue to do and then get frustrated. Some of them may even go to support claiming their TRs don't pass validation. Can it still be fixed somehow?

Or lets approach it from another direction: we have complementing issue #354  When it will be fixed, and this list will be displayed without need to click "Search" button, will it also resolve this issue? If it will, we may just wait for that moment.

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

And let's also consider next situation. Some customers have enormous list of TRs, dozens of entries. So, admin puts some string to this search field that will return him just one TR he needs to update. He updates it with new metadata and clicks "Update", then opens list of TRs, puts **the very same string** to search field (what is simply logical step) - and I suppose he'll never see his TR updated too, no matter how long he'll be hitting this "Search" button? That doesn't seem right..

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

I've tested case I described, and it seems if any string is put into search field, it updates entries successfully. So it's only becomes an issue when search is done with empty field, as you said. Still, there will be also cases, when user has not so much TRs, and he won't see a reason to use some search filter to list them. After the issue #354  will be fixed, we'll need to also make sure simply refreshing the page will update all states of TRs in the list. Is it possible?

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

No, I'm sorry, about that issue - it exists, works the way I described it, I retested it now properly. There is strange detail though - "Validation Status" is being updated, but "Status" - is not, until I leave to some other page, then return. We need to fix this, after all, seem like the search feature needs to be redesigned a bit

**Comment by shekhar16** [comment]
@aliaksander-samuseu The functionality to load same list as old on same search string is implemented earlier may be to improve performance as user is searching same string,
As per the code, if old string = new string will not hit ldap ,it will show existing list.

**Comment by shekhar16** [comment]
@aliaksander-samuseu if you need to hit every time on ldap on every search button click, please let me know.I will do that.
and the issue in https://github.com/GluuFederation/oxTrust/issues/354  was related to initial page when user lands on this page the list was not visible.so we make list displayed from initial view ,user doesn't have to search initially.

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

> the issue in #354 was related to initial page when user lands on this page the list was not visible.so we make list displayed from initial view ,user doesn't have to search initially.

Yes, I understand this. But let's imagine it was implemented. And also this time lets consider case when user doesn't have much TRs in the system, there are all fit into the initial page. So when user will be adding/updating some TR, he'll either be redirected by app, or will go by himself to that page with the list of TRs. As it takes some time to TR to be validated, he'll be trying to refresh this list, either by reloading the whole page, or by hitting the search button. Please note there is no reason to him to put something into the search field at all, so 95% of users won't do that, that's not intuitive. Will he able to see updates to its status in this case? If search feature works as you described, I guess he won't.

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

>The functionality to load same list as old on same search string is implemented earlier may be to improve performance as user is searching same string,
>As per the code, if old string = new string will not hit ldap ,it will show existing list.

I still see at as a problem. Let's return again to case when there are a lot of TRs, and they take several pages. So when admin open the "Trust relationships" page initially, he can't see TR he needs at all. So he **must** either use pagination controls, or, more likely, the search feature to find his TR. After he's updated it, he most likely will use the same string again to search for it, and will try to click "Search" button with this string again, waiting when it's got validates and set to active. Changing this string each time he does the search is again not intuitive, most of users won't do that.

> if you need to hit every time on ldap on every search button click, please let me know

That may be excessive, I understand your point that initial design most likely was created to off-load LDAP directory. But how about that: we could add some sort of "cool-down" element, and in case the same search string is used as in previous search (including an empty string), and 10,20,30 seconds hasn't passed yet, oxTrust will return the same result as before. But if 10,20,30 seconds has passed from that previous search, it will do actual LDAP search again and provide current result. In case the search string has been changed it will do the actual search right away, so it will only apply to cases when the same search string is used again and again.

Does it make sense?

**Comment by aliaksander-samuseu** [comment]
@shekhar16 

>in case the same search string is used as in previous search (including an empty string), and 10,20,30 seconds hasn't passed yet, oxTrust will return the same result as before. But if 10,20,30 seconds has passed from that previous search, it will do actual LDAP search again and provide current result

Or probably we could go with some simple counter. I.e., if search string isn't changed, it will do actual LDAP search only at 5th,10th,15th click to the "Search" button. Whatever is easier to implement.

**Comment by shekhar16** [comment]
Fixed.
@aliaksander-samuseu  The list will now reloaded after every search click and refresh automatically aswel.

**Comment by nynymike** [comment]
@shekhar16 Is this closed?
